cmd/skopeo/sync.go

```diff
@@ -422,6 +424,71 @@ func imagesToCopyFromRegistry(registryName string, cfg registrySyncConfig, sourc
 			Context:   serverCtx})
 	}
 
+	// handle tags matching semver
+	for imageName, semverConstraints := range cfg.ImagesBySemver {
```

Comment:
This is really quite a bit of duplicated code between (`cfg.Images`,),  `cfg.ImagesByTagRegex` and `cfg.ImagesBySemver`, and the function is now almost 200 lines.

Ignoring `cfg.Images`, please split the shared part, listing tags + filtering based on a predicate, into a separate function, parametrized by a filtering predicate; broadly similar to what #965 did.

(Alternatively, it would make sense to process each repo once, i.e. to extract the list of repos from all three `cfg.Images*` options, and then to iterate per-repo, implementing all three options in an inner loop. But I think that’s a larger refactoring than necessary for _this_ PR, and unfair to impose on this PR as a requirement.)

Reply:
I've attempted to do just that with my latest commits.